crates/types/src/hydration.rs

```diff
@@ -318,6 +343,99 @@ impl DehydratedBidSubmissionFulu {
+                // hashed transaction, hydrate it
+                let bytes = tx.as_ref().try_into().unwrap();
+                let hash = u64::from_le_bytes(bytes);
+                let Some(cached_tx) = order_cache.transactions.get(&hash) else {
+                    last_err = Err(HydrationError::UnknownTxHash { index });
+                    continue;
+                };
+
+                tx_cache_hits += 1;
+                *tx = cached_tx.clone();
+            } else {
+                /// Max size of the rlp encoded tx sig (v 1 byte, s,r 32 bytes each with a leading
+                /// rlp length prefix)
+                const TX_KEY_SIZE: usize = 67;
+
+                if tx.len() < TX_KEY_SIZE {
+                    last_err = Err(HydrationError::InvalidTxLength { length: tx.len(), index });
+                    continue;
+                }
+
+                let mut hasher = FxHasher::default();
+                let last_slice = &tx[tx.len() - TX_KEY_SIZE..];
+                hasher.write(last_slice);
+                let hash = hasher.finish();
+                order_cache.transactions.insert(hash, tx.clone());
+            };
+        }
+
+        // hydrate blobs
+
+        let mut blob_cache_hits: usize = 0;
+        let new_blobs = self.blobs_bundle.new_items.len();
+        for blob_item in self.blobs_bundle.new_items {
+            order_cache.blobs_fulu.insert(blob_item.commitment, (blob_item.proof, blob_item.blob));
+        }
+
+        if self.blobs_bundle.commitments.len() > max_blobs_per_block {
+            last_err = Err(HydrationError::TooManyBlobs {
+                blobs: self.blobs_bundle.commitments.len(),
+                max: max_blobs_per_block,
+            });
+        }
+
+        last_err?;
+
+        let mut sidecar = BlobsBundleV2::with_capacity(self.blobs_bundle.commitments.len());
+        for (index, commitment) in self.blobs_bundle.commitments.into_iter().enumerate() {
+            let Some((proofs, blob)) = order_cache.blobs_fulu.get(&commitment) else {
+                return Err(HydrationError::UnknownBlobHashFulu { commitment, index });
+            };
+
+            // safe because we checked the length above
+            sidecar.commitments.push(commitment).unwrap();
+            for proof in proofs {
+                sidecar.proofs.push(*proof);
+            }
+            sidecar.blobs.push(blob.clone());
+            blob_cache_hits += 1;
+        }
+
+        blob_cache_hits = blob_cache_hits.saturating_sub(new_blobs);
+
+        let submission = SignedBidSubmission::Fulu(SignedBidSubmissionFulu {
+            message: self.message,
+            execution_payload: Arc::new(self.execution_payload),
+            blobs_bundle: Arc::new(BlobsBundle::V2(sidecar)),
+            execution_requests: self.execution_requests,
+            signature: self.signature,
+        });
+
+        Ok(HydratedData { submission, tx_cache_hits, blob_cache_hits, tx_root: self.tx_root })
```

Comment:
does this duplicate a function in `DehydratedBidSubmission`? It would be preferable not to duplicate the function so we don't have to remember to make future changes in 2 places.

Can we share the logic somehow? Perhaps define a trait that both structs implement? e.g. trait with method to access payload, blobs_bundle etc.?